status_test.go

```diff
@@ -0,0 +1,193 @@
+)
+
+var _ = Describe("Cluster Operator status controller", func() {
+	var osClient *osclientset.Clientset
+	var statusController *statusController
+	var stop chan struct{}
+
+	BeforeEach(func() {
+		By("Running the controller")
+
+		var err error
+		osClient, err = osclientset.NewForConfig(cfg)
+		Expect(err).NotTo(HaveOccurred())
+
+		stop = make(chan struct{})
+		statusController = NewStatusController(cfg)
+		go func() {
+			defer GinkgoRecover()
+			statusController.Run(1, stop)
+		}()
+	})
+
+	AfterEach(func() {
+		close(stop)
+		os.Unsetenv(releaseVersionEnvVariableName)
+
+		err := osClient.ConfigV1().ClusterOperators().Delete(context.Background(), clusterOperatorName, metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() error {
+			_, err := osClient.ConfigV1().ClusterOperators().Get(context.Background(), clusterOperatorName, metav1.GetOptions{})
+			if err != nil && apierrors.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}, timeout).Should(Succeed())
+	})
+
+	type testCase struct {
+		releaseVersionEnvVariableValue string
+		// Use a function so that runtime information can be populated (eg UID)
+		existingCO *osconfigv1.ClusterOperator
+	}
+
+	DescribeTable("should ensure Cluster Operator status is present",
+		func(tc testCase) {
+
+			expectedVersion := unknownVersionValue
+			if tc.releaseVersionEnvVariableValue != "" {
+				expectedVersion = tc.releaseVersionEnvVariableValue
+
+				Expect(os.Setenv(releaseVersionEnvVariableName, tc.releaseVersionEnvVariableValue)).To(Succeed())
+				statusController.versionGetter.SetVersion(operatorVersionKey, getReleaseVersion())
+			}
+
+			if tc.existingCO != nil {
+				_, err := osClient.ConfigV1().ClusterOperators().Create(context.Background(), tc.existingCO, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+			}
+
+			Eventually(func() bool {
+				co, err := osClient.ConfigV1().ClusterOperators().Get(context.Background(), clusterOperatorName, metav1.GetOptions{})
+				if err != nil {
+					return false
+				}
+
+				// check version.
+				if len(co.Status.Versions) != 1 {
+					return false
+				}
+				if co.Status.Versions[0].Name != operatorVersionKey && co.Status.Versions[0].Version != expectedVersion {
+					return false
+				}
+
+				// check conditions.
+				if v1helpers.IsStatusConditionFalse(co.Status.Conditions, osconfigv1.OperatorAvailable) {
+					return false
+				}
+				if v1helpers.IsStatusConditionFalse(co.Status.Conditions, osconfigv1.OperatorUpgradeable) {
+					return false
+				}
+				if v1helpers.IsStatusConditionTrue(co.Status.Conditions, osconfigv1.OperatorDegraded) {
+					return false
+				}
+				if v1helpers.IsStatusConditionTrue(co.Status.Conditions, osconfigv1.OperatorProgressing) {
+					return false
+				}
+
+				// check related objects.
+				if !equality.Semantic.DeepEqual(co.Status.RelatedObjects, relatedObjects) {
+					return false
+				}
+				return true
+			}, timeout).Should(BeTrue())
```

Comment:
We only need to check that one sync has been successful right? I wonder if this might be simpler if we have a smaller eventually then multiple assertions after, this way around the assertions should be easier to interpret and work out which part of the code failed

```suggestion
			var co *osconfigv1.ClusterOperator
			Eventually(func() (bool, err) {
				var err error
				co, err = osClient.ConfigV1().ClusterOperators().Get(context.Background(), clusterOperatorName, metav1.GetOptions{})
				if err != nil {
					return false, err
				}
				// Successful sync means CO exists and the status is not empty
				return co.Status != osconfigv1.ClusterStatusOperator{}, nil
			}, timeout).Should(BeTrue())
			
			// check version.
			Expect(co.Status.Versions).To(HaveLen(1))
			Expect(co.Status.Versions[0].Name).To(Equal(operatorVersionKey))
			Expect(co.Status.Versions[0].Version).To(Equal(expectedVersion))
			
			// check conditions.
			Expect(v1helpers.IsStatusConditionTrue(co.Status.Conditions, osconfigv1.OperatorAvailable)).To(BeTrue())
			Expect(v1helpers.IsStatusConditionTrue(co.Status.Conditions, osconfigv1.OperatorUpgradeable)).To(BeTrue())
			Expect(v1helpers.IsStatusConditionFalse(co.Status.Conditions, osconfigv1.OperatorDegraded)).To(BeTrue())
			Expect(v1helpers.IsStatusConditionFalse(co.Status.Conditions, osconfigv1.OperatorProgressing)).To(BeTrue())
			
			// check related objects.
			Expect(co.Status.RelatedObjects).To(Equal(relatedObjects))
```

Reply:
mm I'd like that but I don't think we can though. We first create the object, then reconcile status towards expected relatedObjects, then reconcile status towards expected conditions.
There's scenarios where the object and status exists so eventually would succeed but we might be still reconciling status, right?

Reply:
Ahh yes, I see we are making two calls to update status, this wont work right now, I'm not too fussed either way tbh, I don't think changing the code to fit the tests makes sense 😅 

Does make me wonder though, should the related objects be in `syncStatus` rather than in the create? What's the benefit of having it there?

I'm concerned that calling update status multiple times might cause 409 conflict responses from the API when reconciling, have you seen that at all?

Reply:
>Does make me wonder though, should the related objects be in syncStatus rather than in the create? What's the benefit of having it there?
I'm concerned that calling update status multiple times might cause 409 conflict responses from the API when reconciling, have you seen that at all?

Make sense, let me do that.

Reply:
Did you try this suggestion? I think it might work now the status has been refactored? WDYT?